### PR TITLE
Fix utilization of record offset in procedure calls, refactor outputs

### LINKED ISSUE #1701: Run BFS query crash redis
Hi ,we encounter the other crash again, this could be reproduced with the mock data,thanks in advance.

this is the crash detail: [crash_detail.txt](https://github.com/RedisGraph/RedisGraph/files/6471400/crash_detail.txt)

this is mock data:  [bfs_crash.zip](https://github.com/RedisGraph/RedisGraph/files/6471407/bfs_crash.zip)

1) docker run -p 6379:6379 -it --rm redislabs/redisgraph

2) redisgraph-bulk-loader GRAPH_DEMO  --enforce-schema -p 6379  -n /tmp/bfs_crash/nodes/element.csv -n /tmp/bfs_crash/nodes/fact.csv -r /tmp/bfs_crash/rel/FormOf.csv -r /tmp/bfs_crash/rel/IsA.csv -r /tmp/bfs_crash/rel/IsElementOf.csv -r /tmp/bfs_crash/rel/RelatedTo.csv -r /tmp/bfs_crash/rel/RootOf.csv -r /tmp/bfs_crash/rel/Synonym.csv

3) GRAPH.QUERY "GRAPH_DEMO" "MATCH (n0:element {name: 'event loop'}) CALL algo.BFS(n0, 2, NULL) YIELD nodes as ns0,edges as e0  MATCH (n1:element {name: 'condition'}) CALL algo.BFS(n1, 2, NULL) YIELD nodes as ns1,edges as e1    MATCH (n2:element {name: 'construct'}) CALL algo.BFS(n2, 2, NULL) YIELD nodes as ns2,edges as e2      RETURN ns0,e0,ns1,e1,ns2,e2"

Run the query in the step 3 will crash redis.

by the way, the other issue is if we choice to return " RETURN size(ns0),size(ns1),size(ns2) " in the step 3, the result seem not right.

### PR DESCRIPTION
**Pull request by jeffreylovitz**
Resolves #1701 

Previously, the default variable names for `YIELD` expressions were placed in the record map and any aliases applied to them were linked to them using `OpBase_AliasModifier`. This function increases the size of the record map without introducing new members, introducing empty spaces in the record, and should be avoided. Additionally, mapping the default variable names causes collisions when the same procedure is called twice in one query, in the case of `nodes` on the linked issue. This causes both aliases to be mapped to the first `nodes` output and the second BFS call to never have its outputs checked. This led to a double free when emitting results.

This PR also standardizes and simplifies the way procedures register their outputs. Non-yielded results are not computed, only values are stored rather than key-value pairs with output names, and other changes have been made.

### CODE REVIEW

**Review by swilly22 — CHANGES_REQUESTED** [review]
This is a welcome change!
Please see additional comments

File: src/procedures/proc_bfs.c [hunk 1]
```diff
@@ -33,49 +33,29 @@ typedef struct {
...
-	if(yield_edges) {
-		array_append(ctx->output, SI_ConstStringVal("edges"));
-		ctx->edges_output_idx = array_len(ctx->output);
-		array_append(ctx->output, SI_NullVal()); // Place holder.
+		if(strcasecmp("edges", yield[i]) == 0) {
+			ctx->yield_edges = ctx->output + idx;
+			idx ++;
+			continue;
+		}
 	}
```
[the remark this answers is not in the document]
Comment (jeffreylovitz) [reply src/procedures/proc_bfs.c:63]: That's an impossible situation. If the YIELD is omitted both will be yielded, and if YIELD is included AST validations will make sure there is a valid argument.

File: src/procedures/proc_bfs.c [hunk 2]
```diff
@@ -119,7 +99,7 @@ static ProcedureResult Proc_BFS_Invoke(ProcedureCtx *ctx,
 	} else {
 		Schema *s = GraphContext_GetSchema(gc, reltype, SCHEMA_EDGE);
 		// failed to find schema, first step will return NULL
-		if(!s) return PROCEDURE_OK; 
+		if(!s) return PROCEDURE_OK;
 
 		bfs_ctx->reltype_id = s->id;
 		RG_Matrix_export(&R, Graph_GetRelationMatrix(gc->g, s->id, false));
```
[the remark this answers is not in the document]
Comment (jeffreylovitz) [reply src/procedures/proc_bfs.c:114]: `SIArray_Append` clones the input `SIValue`, so I think this is a safe usage. Is there something I'm missing?

File: src/procedures/proc_fulltext_query.c [hunk 3]
```diff
@@ -26,9 +26,30 @@ typedef struct {
...
+			ctx->yield_node = ctx->output + idx;
+			idx++;
+			continue;
+		}
+		if(strcasecmp("score", yield[i]) == 0) {
+			ctx->yield_score = ctx->output + idx;
+			idx++;
+			continue;
+		}
+	}
```
[the remark this answers is not in the document]
Comment (jeffreylovitz) [reply src/procedures/proc_fulltext_query.c:54]: https://github.com/RedisGraph/RedisGraph/pull/2028#discussion_r751512215

File: src/procedures/proc_pagerank.c [hunk 4]
```diff
@@ -25,9 +25,28 @@ typedef struct {
...
+			idx++;
+			continue;
+		}
+
+		if(strcasecmp("score", yield[i]) == 0) {
+			ctx->yield_score = ctx->output + idx;
+			idx++;
+			continue;
+		}
+	}
```
[the remark this answers is not in the document]
Comment (jeffreylovitz) [reply src/procedures/proc_pagerank.c:51]: https://github.com/RedisGraph/RedisGraph/pull/2028#discussion_r751512215

File: src/procedures/proc_procedures.c [hunk 5]
```diff
@@ -17,10 +17,30 @@ extern rax *__procedures;
...
+		if(strcasecmp("name", yield[i]) == 0) {
+			ctx->yield_name = ctx->output + idx;
+			idx++;
+			continue;
+		}
+		if(strcasecmp("mode", yield[i]) == 0) {
+			ctx->yield_mode = ctx->output + idx;
+			idx++;
+			continue;
+		}
```
[the remark this answers is not in the document]
Comment (jeffreylovitz) [reply src/procedures/proc_procedures.c:42]: In Neo4j this is possible, but for us, repeating an output causes an AST validation error:
```
redis-cli G "CALL dbms.procedures YIELD name AS a, name AS b"
(error) Variable `name` already declared
```
This isn't really the correct behavior, but I'm not sure if we want to fix it, as then we'll have to deal with the problem you describe.

File: src/procedures/proc_bfs.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (swilly22) [reply src/procedures/proc_bfs.c:63]: 👍

File: src/procedures/proc_bfs.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (swilly22) [reply src/procedures/proc_bfs.c:114]: yeah but aren't we copying a pointer, which later gets modified?

File: src/procedures/proc_procedures.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (swilly22) [reply src/procedures/proc_procedures.c:42]: let's keep it as is

File: src/procedures/proc_bfs.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (jeffreylovitz) [reply src/procedures/proc_bfs.c:114]: For SIValues of type `node` or `edge`, we memcpy the memory region - https://github.com/RedisGraph/RedisGraph/blob/master/src/value.c#L153-L169

**Review by swilly22 — APPROVED** [review]

**Pull request merged**